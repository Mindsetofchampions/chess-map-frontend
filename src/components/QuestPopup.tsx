import React, { useEffect } from 'react';

type Difficulty = 'easy' | 'medium' | 'hard';

export interface Quest {
  id: string;
  title: string;
  description: string;
  coins: number;
  location?: string;
  estimatedTime?: string;
  difficulty?: Difficulty;
}

export interface QuestPopupProps {
  quest: Quest | null;
  isOpen: boolean;
  onClose: () => void;
  onStartQuest: (id: string) => void;
  position?: { x: number; y: number };
}

const difficultyLabel: Record<Difficulty, string> = {
  easy: 'Easy',
  medium: 'Medium',
  hard: 'Hard',
};

const difficultyClass: Record<Difficulty, string> = {
  easy: 'text-cyber-green-400',
  medium: 'text-yellow-400',
  hard: 'text-red-400',
};

const QuestPopup: React.FC<QuestPopupProps> = ({
  quest,
  isOpen,
  onClose,
  onStartQuest,
  position,
}) => {
  useEffect(() => {
    document.body.style.overflow = isOpen ? 'hidden' : 'unset';
    return () => {
      document.body.style.overflow = 'unset';
    };
  }, [isOpen]);

  if (!isOpen || !quest) return null;

  const difficulty = quest.difficulty ?? 'medium';

  const style: React.CSSProperties | undefined = position
    ? { position: 'fixed', left: `${position.x}px`, top: `${position.y}px` }
    : undefined;

  const handleStart = () => {
    onStartQuest(quest.id);
    onClose();
  };

  return (
    <div className='fixed inset-0 z-50 flex items-center justify-center'>
      <div
        role='dialog'
        aria-modal='true'
        aria-labelledby='quest-title'
        aria-describedby='quest-description'
        className='bg-white/10 backdrop-blur-xl border rounded-2xl shadow-xl p-4 max-w-md w-full text-white relative'
        // Close only via explicit buttons to satisfy a11y rules
        style={style}
      >
        <button
          aria-label='Close quest popup'
          className='absolute right-2 top-2 p-1 rounded-full'
          onClick={onClose}
        >
          Ã—
        </button>

        <h2 id='quest-title' className='text-xl font-bold mb-2'>
          {quest.title}
        </h2>

        <p id='quest-description' className='text-sm mb-3'>
          {quest.description}
        </p>

        <div className='grid grid-cols-3 gap-2 text-sm mb-4'>
          <div>
            <div className='opacity-80'>Coins</div>
            <div>{quest.coins}</div>
          </div>
          {quest.location && (
            <div>
              <div className='opacity-80'>Location</div>
              <div>{quest.location}</div>
            </div>
          )}
          {quest.estimatedTime && (
            <div>
              <div className='opacity-80'>Time</div>
              <div>{quest.estimatedTime}</div>
            </div>
          )}
          <div>
            <div className='opacity-80'>Difficulty</div>
            <div className={difficultyClass[difficulty]}>{difficultyLabel[difficulty]}</div>
          </div>
        </div>

        <div className='flex gap-2'>
          <button className='border rounded-xl px-4 py-2' onClick={handleStart}>
            Start Quest
          </button>
          <button className='border rounded-xl px-4 py-2' onClick={onClose}>
            Cancel
          </button>
        </div>
      </div>
    </div>
  );
};

export default QuestPopup;
