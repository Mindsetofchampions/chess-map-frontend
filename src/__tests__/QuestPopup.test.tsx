import { render, screen, fireEvent } from '@testing-library/react';
import React from 'react';

import '@testing-library/jest-dom';
import QuestPopup from '../components/QuestPopup';

/**
 * Test suite for QuestPopup component
 *
 * Tests comprehensive popup functionality including:
 * - Rendering with quest data
 * - Modal behavior and interactions
 * - Accessibility compliance
 * - Animation states
 * - Button functionality
 */
describe('QuestPopup', () => {
  const mockOnClose = jest.fn();
  const mockOnStartQuest = jest.fn();

  const mockQuest = {
    id: 'test-quest-1',
    title: 'Character Development Challenge',
    description:
      'Complete wisdom-building activities with Hootie the Owl to develop strong character traits.',
    coins: 50,
    location: 'Central Park',
    estimatedTime: '15 minutes',
    difficulty: 'medium' as const,
  };

  const defaultProps = {
    quest: mockQuest,
    isOpen: true,
    onClose: mockOnClose,
    onStartQuest: mockOnStartQuest,
  };

  beforeEach(() => {
    jest.clearAllMocks();
    // Reset body overflow style
    document.body.style.overflow = 'unset';
  });

  /**
   * Basic rendering tests
   */
  it('renders quest popup when open', () => {
    render(<QuestPopup {...defaultProps} />);

    expect(screen.getByRole('dialog')).toBeInTheDocument();
    expect(screen.getByText('Character Development Challenge')).toBeInTheDocument();
    expect(screen.getByText(/Complete wisdom-building activities/)).toBeInTheDocument();
  });

  it('does not render when closed', () => {
    render(<QuestPopup {...defaultProps} isOpen={false} />);

    expect(screen.queryByRole('dialog')).not.toBeInTheDocument();
  });

  it('renders null when no quest provided', () => {
    render(<QuestPopup {...defaultProps} quest={null} />);

    expect(screen.queryByRole('dialog')).not.toBeInTheDocument();
  });

  /**
   * Quest information display tests
   */
  it('displays quest metadata correctly', () => {
    render(<QuestPopup {...defaultProps} />);

    // Check coins display
    expect(screen.getByText('50')).toBeInTheDocument();

    // Check location
    expect(screen.getByText('Central Park')).toBeInTheDocument();

    // Check estimated time
    expect(screen.getByText('15 minutes')).toBeInTheDocument();

    // Check difficulty badge
    expect(screen.getByText('Medium')).toBeInTheDocument();
  });

  it('handles optional quest properties gracefully', () => {
    const minimalQuest = {
      id: 'minimal-quest',
      title: 'Minimal Quest',
      description: 'A quest with minimal information',
      coins: 25,
    };

    render(<QuestPopup {...defaultProps} quest={minimalQuest} />);

    expect(screen.getByText('Minimal Quest')).toBeInTheDocument();
    expect(screen.getByText('25')).toBeInTheDocument();
    expect(screen.queryByText('Central Park')).not.toBeInTheDocument();
  });

  /**
   * Interaction tests
   */
  it('calls onClose when close button is clicked', () => {
    render(<QuestPopup {...defaultProps} />);

    const closeButton = screen.getByLabelText('Close quest popup');
    fireEvent.click(closeButton);

    expect(mockOnClose).toHaveBeenCalledTimes(1);
  });

  it('calls onStartQuest when start button is clicked', () => {
    render(<QuestPopup {...defaultProps} />);

    const startButton = screen.getByText('Start Quest');
    fireEvent.click(startButton);

    expect(mockOnStartQuest).toHaveBeenCalledWith('test-quest-1');
    expect(mockOnClose).toHaveBeenCalledTimes(1);
  });

  it('calls onClose when cancel button is clicked', () => {
    render(<QuestPopup {...defaultProps} />);

    const cancelButton = screen.getByText('Cancel');
    fireEvent.click(cancelButton);

    expect(mockOnClose).toHaveBeenCalledTimes(1);
  });

  it('calls onClose when backdrop is clicked', () => {
    render(<QuestPopup {...defaultProps} />);

    const backdrop = screen.getByRole('dialog').parentElement;
    fireEvent.click(backdrop!);

    expect(mockOnClose).toHaveBeenCalledTimes(1);
  });

  it('does not close when clicking inside popup content', () => {
    render(<QuestPopup {...defaultProps} />);

    const popupContent = screen.getByRole('dialog');
    fireEvent.click(popupContent);

    expect(mockOnClose).not.toHaveBeenCalled();
  });

  /**
   * Accessibility tests
   */
  it('has proper ARIA attributes', () => {
    render(<QuestPopup {...defaultProps} />);

    const dialog = screen.getByRole('dialog');
    expect(dialog).toHaveAttribute('aria-modal', 'true');
    expect(dialog).toHaveAttribute('aria-labelledby', 'quest-title');
    expect(dialog).toHaveAttribute('aria-describedby', 'quest-description');
  });

  it('manages body scroll correctly', () => {
    const { rerender } = render(<QuestPopup {...defaultProps} />);

    // Should prevent body scroll when open
    expect(document.body.style.overflow).toBe('hidden');

    // Should restore body scroll when closed
    rerender(<QuestPopup {...defaultProps} isOpen={false} />);
    expect(document.body.style.overflow).toBe('unset');
  });

  /**
   * Difficulty color tests
   */
  it('applies correct difficulty colors', () => {
    const { rerender } = render(<QuestPopup {...defaultProps} />);

    // Medium difficulty
    expect(screen.getByText('Medium')).toHaveClass('text-yellow-400');

    // Easy difficulty
    rerender(<QuestPopup {...defaultProps} quest={{ ...mockQuest, difficulty: 'easy' }} />);
    expect(screen.getByText('Easy')).toHaveClass('text-cyber-green-400');

    // Hard difficulty
    rerender(<QuestPopup {...defaultProps} quest={{ ...mockQuest, difficulty: 'hard' }} />);
    expect(screen.getByText('Hard')).toHaveClass('text-red-400');
  });

  /**
   * Position-based rendering tests
   */
  it('applies custom position styling when provided', () => {
    const position = { x: 100, y: 200 };
    render(<QuestPopup {...defaultProps} position={position} />);

    const popup = screen.getByRole('dialog');
    // Position styling is applied via inline styles
    expect(popup).toHaveStyle({
      position: 'fixed',
      left: '100px',
      top: '200px',
    });
  });

  /**
   * Snapshot tests for visual regression
   */
  describe('Snapshot Tests', () => {
    it('matches snapshot for full quest data', () => {
      const { container } = render(<QuestPopup {...defaultProps} />);
      expect(container.firstChild).toMatchSnapshot();
    });

    it('matches snapshot for minimal quest data', () => {
      const minimalQuest = {
        id: 'minimal',
        title: 'Simple Quest',
        description: 'Basic quest description',
        coins: 10,
      };

      const { container } = render(<QuestPopup {...defaultProps} quest={minimalQuest} />);
      expect(container.firstChild).toMatchSnapshot();
    });

    it('matches snapshot for different difficulty levels', () => {
      const easyQuest = { ...mockQuest, difficulty: 'easy' as const };
      const { container } = render(<QuestPopup {...defaultProps} quest={easyQuest} />);
      expect(container.firstChild).toMatchSnapshot();
    });

    it('matches snapshot when closed', () => {
      const { container } = render(<QuestPopup {...defaultProps} isOpen={false} />);
      expect(container.firstChild).toMatchSnapshot();
    });
  });
});
