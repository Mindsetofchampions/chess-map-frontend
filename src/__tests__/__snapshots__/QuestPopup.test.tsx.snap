// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`QuestPopup Snapshot Tests matches snapshot for different difficulty levels 1`] = `
<div
  class="fixed inset-0 z-50 flex items-center justify-center"
>
  <div
    aria-describedby="quest-description"
    aria-labelledby="quest-title"
    aria-modal="true"
    class="bg-white/10 backdrop-blur-xl border rounded-2xl shadow-xl p-4 max-w-md w-full text-white relative"
    role="dialog"
  >
    <button
      aria-label="Close quest popup"
      class="absolute right-2 top-2 p-1 rounded-full"
    >
      ×
    </button>
    <h2
      class="text-xl font-bold mb-2"
      id="quest-title"
    >
      Character Development Challenge
    </h2>
    <p
      class="text-sm mb-3"
      id="quest-description"
    >
      Complete wisdom-building activities with Hootie the Owl to develop strong character traits.
    </p>
    <div
      class="grid grid-cols-3 gap-2 text-sm mb-4"
    >
      <div>
        <div
          class="opacity-80"
        >
          Coins
        </div>
        <div>
          50
        </div>
      </div>
      <div>
        <div
          class="opacity-80"
        >
          Location
        </div>
        <div>
          Central Park
        </div>
      </div>
      <div>
        <div
          class="opacity-80"
        >
          Time
        </div>
        <div>
          15 minutes
        </div>
      </div>
      <div>
        <div
          class="opacity-80"
        >
          Difficulty
        </div>
        <div
          class="text-cyber-green-400"
        >
          Easy
        </div>
      </div>
    </div>
    <div
      class="flex gap-2"
    >
      <button
        class="border rounded-xl px-4 py-2"
      >
        Start Quest
      </button>
      <button
        class="border rounded-xl px-4 py-2"
      >
        Cancel
      </button>
    </div>
  </div>
</div>
`;

exports[`QuestPopup Snapshot Tests matches snapshot for full quest data 1`] = `
<div
  class="fixed inset-0 z-50 flex items-center justify-center"
>
  <div
    aria-describedby="quest-description"
    aria-labelledby="quest-title"
    aria-modal="true"
    class="bg-white/10 backdrop-blur-xl border rounded-2xl shadow-xl p-4 max-w-md w-full text-white relative"
    role="dialog"
  >
    <button
      aria-label="Close quest popup"
      class="absolute right-2 top-2 p-1 rounded-full"
    >
      ×
    </button>
    <h2
      class="text-xl font-bold mb-2"
      id="quest-title"
    >
      Character Development Challenge
    </h2>
    <p
      class="text-sm mb-3"
      id="quest-description"
    >
      Complete wisdom-building activities with Hootie the Owl to develop strong character traits.
    </p>
    <div
      class="grid grid-cols-3 gap-2 text-sm mb-4"
    >
      <div>
        <div
          class="opacity-80"
        >
          Coins
        </div>
        <div>
          50
        </div>
      </div>
      <div>
        <div
          class="opacity-80"
        >
          Location
        </div>
        <div>
          Central Park
        </div>
      </div>
      <div>
        <div
          class="opacity-80"
        >
          Time
        </div>
        <div>
          15 minutes
        </div>
      </div>
      <div>
        <div
          class="opacity-80"
        >
          Difficulty
        </div>
        <div
          class="text-yellow-400"
        >
          Medium
        </div>
      </div>
    </div>
    <div
      class="flex gap-2"
    >
      <button
        class="border rounded-xl px-4 py-2"
      >
        Start Quest
      </button>
      <button
        class="border rounded-xl px-4 py-2"
      >
        Cancel
      </button>
    </div>
  </div>
</div>
`;

exports[`QuestPopup Snapshot Tests matches snapshot for minimal quest data 1`] = `
<div
  class="fixed inset-0 z-50 flex items-center justify-center"
>
  <div
    aria-describedby="quest-description"
    aria-labelledby="quest-title"
    aria-modal="true"
    class="bg-white/10 backdrop-blur-xl border rounded-2xl shadow-xl p-4 max-w-md w-full text-white relative"
    role="dialog"
  >
    <button
      aria-label="Close quest popup"
      class="absolute right-2 top-2 p-1 rounded-full"
    >
      ×
    </button>
    <h2
      class="text-xl font-bold mb-2"
      id="quest-title"
    >
      Simple Quest
    </h2>
    <p
      class="text-sm mb-3"
      id="quest-description"
    >
      Basic quest description
    </p>
    <div
      class="grid grid-cols-3 gap-2 text-sm mb-4"
    >
      <div>
        <div
          class="opacity-80"
        >
          Coins
        </div>
        <div>
          10
        </div>
      </div>
      <div>
        <div
          class="opacity-80"
        >
          Difficulty
        </div>
        <div
          class="text-yellow-400"
        >
          Medium
        </div>
      </div>
    </div>
    <div
      class="flex gap-2"
    >
      <button
        class="border rounded-xl px-4 py-2"
      >
        Start Quest
      </button>
      <button
        class="border rounded-xl px-4 py-2"
      >
        Cancel
      </button>
    </div>
  </div>
</div>
`;

exports[`QuestPopup Snapshot Tests matches snapshot when closed 1`] = `null`;
